### ISSUE TITLE
[HOLD for payment 2023-08-30] [$1000] Navigate back to public room after clicking on Sign-in as unauthenticated user

### ISSUE THREAD

**Issue by marcochavezf** [Expensify/App#20558]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Open a public room as unauthenticated user (i.e. https://staging.new.expensify.com/r/8813851443373010)
2. Click on sign-in 
3. Click on the back button

https://github.com/Expensify/App/assets/6829422/4958ecad-0ed8-48a4-86f3-37b5960ed48c

## Expected Result:
It should redirect you to the public room

## Actual Result:
It just redirects you the previous browser page

## Workaround:
Click again on the link

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android / native
- [x] Android / Chrome
- [ ] iOS / native
- [x] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** v1.3.25-8
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** 
**Expensify/Expensify Issue URL:** 
**Issue reported by:**
**Slack conversation:**

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~017a21cdc86c5eee7f</li>
        <li>Upwork Job ID: 1667332458840076288</li>
        <li>Last Price Increase: 2023-06-10</li>
    </ul>
</details>

**Comment by marcochavezf** [comment]
Making it external in case someone wants to propose a solution during the weekend. Just some additional info, we're setting the reportID of the public room [here](https://github.com/Expensify/App/blob/d5374be5435eea4b94978f93588ebad8ffdc0867/src/libs/actions/Session/index.js#L96) in Onyx when we "log out" the anonymous user (maybe we can navigate to that reportID when the user clicks on the back button).

**Comment by daordonez11** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
When accessing a public room as an anonymous user and then clicking on sign in the navigation stack is lost

### What is the root cause of that problem?
When sign in is pressed signOutAndRedirectToSignIn method is called, finally calling clearStorageAndRedirect.  "removing the authToken redirects the user to the Sign-in page", and cleans all the navigation stack 

### What changes do you think we should make in order to solve the problem?
I would recommend creating an auxiliary method in the Navigation library to "force" the public room in the navigation stack:
`function resetWithLoginAndReport(reportId){
    // if login is accessed after public room include the report in the stack
    navigationRef.current.dispatch(CommonActions.reset({index: 1, routes: [{name: ROUTES.getReportRoute(reportId)},{name: SCREENS.HOME},]}));
}`
 I found "in order to add a history item onto the browser history stack one would need to use the "push" action." [Here](https://github.com/Expensify/App/blob/0e3bbe77a3ed497b49ffb3ee32d952e480bda448/src/libs/Navigation/Navigation.js#L110)

This can be called in two places:
1. In the sign in page connecting the props to onyx, validating if it exists to include it in the browser stack
2. Validate if the value exists in onix on the clean of the session and include it in the stack before navigating to sign in

### What alternative solutions did you explore? (Optional)
First I thought about saving the last link before navigating and cleaning but then I realized that save part was already done and explained by @marcochavezf

**Comment by daordonez11** [comment]
Contributor details
Your Expensify account email: dloz1996@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~01e96b8da58232f028

**Comment by marcochavezf** [comment]
Hi @daordonez11 and thanks for the proposal! Could you share a quick video of that proposal working? Most of the proposal looks good, but I think I'm not understanding completely the "why" of this part:

> I found "in order to add a history item onto the browser history stack one would need to use the "push" action." [Here](https://github.com/Expensify/App/blob/0e3bbe77a3ed497b49ffb3ee32d952e480bda448/src/libs/Navigation/Navigation.js#L110)

**Comment by daordonez11** [comment]
Hey @marcochavezf  for sure working on it! Any specific branch? I'll do it on the commit you shared, yesterday I was working on main. Also quick question, you could save lastOpenesPublicRoom before calling "redirectToSignIn" and including your key in keysToPreserve in the method clearStorageAndRedirect. Any specific reason why it is currently being done after? Sometimes in my test the signin component is loaded before the key exists in onyx so I need to handle updates.

**Comment by daordonez11** [comment]
Hey @marcochavezf here is a video with the fix and will update the proposal 

https://github.com/Expensify/App/assets/13544910/47eb5d8f-e84a-48df-919d-4a919c559d7a

Some important stuff, the original proposal didn't work because PuclicScreens and AuthScreens are different navigators, so you cannot simply navigate between both, hence resetting the navigation doesn't work.
Final solution was to listen navigator popstate, BackHandler didn't work in web app either, and open the report same as deep link does.

**Comment by marcochavezf** [comment]
> you could save lastOpenesPublicRoom before calling "redirectToSignIn" and including your key in keysToPreserve in the method clearStorageAndRedirect. Any specific reason why it is currently being done after?

Oh this was hotfix and is just a temporal solution, since we're still changing things in the backend to share the report with the user signed in. When I implemented it I forgot we could include it in the keysToPerserve, thanks for pointing that!

**Comment by marcochavezf** [comment]
Also thanks for updating the proposal! Yeah, given the current structure of the navigators I think it makes sense to listen the popstate, but I think this solution will work only for web, no? Could you also check a solution for Android native?

**Comment by daordonez11** [comment]
Oh for sure! I focused on a solution that could work on web. For Android Native I tested BackHandler, practically the same solution but different handler (hardwareBackPress). Later when I'm on the PC I can share a video.

**Comment by marcochavezf** [comment]
Sounds good, thanks! I think that will be enough for Android and we can address any detail in the PR. Assigning @daordonez11 🚀

**Comment by daordonez11** [comment]
Thanks @marcochavezf I will apply to upwork then. I expect to have the PR done by tomorrow(June 12) end of day so you can check it on Tuesday, I'll let you know any important update. Also any chance you could help me request the slack access? I sent it last week to the contributors email but still no response.

**Comment by daordonez11** [comment]
Hey @marcochavezf  I already have part of the PR ready but I have 2 questions 
1. How should I manage the code that is only for android and only for web? I'm currently testing with something like this:
```javascript
useEffect(() => {
        if(lastOpenedRoomId && lastOpenedRoomId!==''){
            // Push state is required to trigger the popstate event
            window.history.pushState({}, '', ROUTES.getReportRoute(lastOpenedRoomId));
        }
        const browserBackPressed = () => {
            if(lastOpenedRoomId && lastOpenedRoomId!==''){
                // Use deeplink implementation to change navigation stack
                Report.openReportFromDeepLink(ROUTES.getReportRoute(lastOpenedRoomId), false)
            }else{
                window.history.back();
            }
        };
        const androidBackButtonPressed = () => {
            if(lastOpenedRoomId && lastOpenedRoomId!==''){
                Report.openReportFromDeepLink(ROUTES.getReportRoute(lastOpenedRoomId), false)
            }else{
                BackHandler.exitApp()
            }
          return true;
        };
        // Add event listener for the browser back button press and android back button
        if(typeof window !== 'undefined' && window.addEventListener ) {
            window.addEventListener('popstate', browserBackPressed);
        }
        else {
            const backHandler = BackHandler.addEventListener(
                'hardwareBackPress',
                androidBackButtonPressed,
              );
        }
        
        // Clean up the event listeners when the component unmounts
        return () => {
            if(typeof window !== 'undefined' && window.addEventListener) {
                window.removeEventListener('popstate', browserBackPressed);
            }
            else {
                backHandler.remove();
            }
        };
      }, [lastOpenedRoomId]);
```
But I don't feel all this window ifs are the way of solving it

2. How can I test deeplink in android native? I already have the environment working but I'm not sure how to run the app beginning on the public room

**Comment by marcochavezf** [comment]
@mollfpr was assigned automatically in the PR, so I will assign you also here :)

**Comment by daordonez11** [comment]
Quick summary of the status of the issue: 
With the help of @marcochavezf  and @mountiny, it is decided that the proposal was not enough to solve the RC, hence the scope and reward of the issue is being re-evaluated. Currently, 2 alternatives are being discussed and evaluated with POC

1.  Unifying AuthScreens and PublicScreens to a single RootStack, this enables navigation through the whole application and avoids losing the history stack. This change already has a POC branch [here](https://github.com/daordonez11/DanielExpensify/tree/poc-shared-root) this one is tricky since it might require testing a lot of navigation scenarios and touch the core of the navigation of the app
2. Parting from this issue [18321](https://github.com/Expensify/App/issues/18321) there is already a solution focused on embedding the SignInPage in a modal, the idea would be to reuse this and create a solution that includes Signin in the RHP so the final user can go back.

**Comment by marcochavezf** [comment]
@daordonez11 can you create a WIP PR of the POC? So it will be easier to check/review the changes made so far. And thanks again for the analysis and the effort!

**Comment by daordonez11** [comment]
For sure @marcochavezf  here is the draft with the initial POC code: https://github.com/Expensify/App/pull/20883

**Comment by dylanexpensify** [comment]
Nice one! Looks like we're making good process here

**Comment by daordonez11** [comment]
Update for everyone! After some discussion in https://github.com/Expensify/App/pull/20883 it is found that joining both stacks doesn't quite solve the issue hence it is concluded that the next approach tested should be trying `SignInPage` in the RHP. I will work on a new PR on the weekend with this approach. I expect to have something working for Monday so all of the interested stakeholders can give it a check and test it out.

TLDR; We will start testing option 2 of my last comment

**Comment by daordonez11** [comment]
I have now updated the [branch](https://github.com/daordonez11/DanielExpensify/tree/poc-signin-rhp) and signin modal is shown and working. I'm still trying to figure out a way to make the `SignInPage` adaptive without using `withWindowDimensions` or finding a way to override `isSmallScreenWidth`

**Comment by daordonez11** [comment]
Hey @marcochavezf third is the charm lets hope! Here is the draft PR so you can give it a check, POC of signin modal in RHP https://github.com/Expensify/App/pull/22005

**Comment by marcochavezf** [comment]
Hi @daordonez11, the POC looks good! Let's continue with the RHP sign-in implementation. Just a few issues I noticed:

 1. we're showing the scrollbar in the RHP (we should remove it)
 2. when the user enters the validation code the validation page is still shown 
 3. when the user clicks on one item from the FAB menu, the FAB menu is still open

<img width="930" alt="Screenshot 2023-07-06 at 17 53 13" src="https://github.com/Expensify/App/assets/6829422/45f42747-690a-4493-893d-575e9898689a">

https://github.com/Expensify/App/assets/6829422/e31cd27e-6224-421d-aa41-f2b379e6d9a9

But overall the POC is great!

**Comment by daordonez11** [comment]
Hey Marco! Thanks for the feedback! Will continue working on it then! Also, any feedback from UX? Regarding error #2 it occurs because authToken is cleaned, without that the authentication flow fails. This is automatic due to the logic in AppNavigator I'll think about a workaround on how to avoid it. That is the same reason I created `signInAnonymousAccount` to handle the specific flow of cleaning authTokenType to trigger the auth flow. Maybe with an optimistic flow I don't need to clean it but use the response or a fake token while I receive the response from the server. I'll test and let you know. I'll check the other issues too.

**Comment by mountiny** [comment]
Just noting probably obvious but the header of the RHP should have same colour as the page itself.

Once this is going to be more ready we can ask Shawn our designer for a review of the PR from design perspective.

**Comment by marcochavezf** [comment]
> Regarding error https://github.com/Expensify/App/pull/2 it occurs because authToken is cleaned, without that the authentication flow fails. This is automatic due to the logic in AppNavigator I'll think about a workaround on how to avoid it.

Hi @daordonez11! I was exploring some options to solve this issue and seems there's no easy solution because the app is designed to handle one authToken and I saw how clearing it will display the PublicScreens for a second. I think some possible solutions are:
1. Use the new ClaimAnonymousAccount to change the authToken when the user enters the validation code. But I think we'd need to call some other functions like `App.openApp()` (which is called in AuthScreens) when the new authToken is generated.
2. Display a loading page when the user enters the validation code, but maybe that could be a bit disruptive for the user.

Tomorrow I will continue with some tests with the `ClaimAnonymousAccount`.

**Comment by dylanexpensify** [comment]
@marcochavezf any updates? Also this should have a C+ right?

**Comment by marcochavezf** [comment]
Ah sorry I didn't post an update yesterday, but I'm still working on the `ClaimAnonymousAccount`, I realized that when we're calling `BeginSignIn` we're creating an account for the entered email, so we'd need to merge accounts in `ClaimAnonymousAccount`. Today I will continue with the testing/implementation and post an update here at EOD

For C+, yes once the PR is ready for review (which I think will happen after having the `ClaimAnonymousAccount` command and I estimate it will live next week) we'd need a C+ for review

**Comment by daordonez11** [comment]
I'll work on the other mentioned bugs while Marco works on the backend command. Also https://github.com/Expensify/App/issues/20558#issuecomment-1590347789 I think we already have @mollfpr assigned as C+ !

**Comment by marcochavezf** [comment]
> Also https://github.com/Expensify/App/issues/20558#issuecomment-1590347789 I think we already had @mollfpr assigned as C+ ! For the first PR

Oh that's correct! @mollfpr should be assigned to the final PR, assigning you to the GH issue

**Comment by marcochavezf** [comment]
I was playing with the new command (and removing this part of the [code](https://github.com/Expensify/App/blob/654651ff21ead895f72af76c8fba0f48d5f30816/src/libs/actions/Session/index.js#L892-L899) to avoid removing the `authToken`) and the functionality looks good so far. We'd need a way to hide the RHP when the command resolves (I think we'd need to send a new onyx variable from the server to do this):

https://github.com/Expensify/App/assets/6829422/f16f4a46-cc5b-4057-ae2f-68e5fe0c6b05

**Comment by marcochavezf** [comment]
Hi @daordonez11, I've been making progress with the command, and I realized that we will keep using the `Session.signIn` function. The app will automatically send the authToken of the anonymous account, and we'll do the "claiming" account in the server if the anonymous authToken is present. 

One additional thing that we'd need to do in the App is close the RHP when we insert the new user data into Onyx. I imagine it could be something like this:

```js
Onyx.connect({
    key: ONYXKEYS.SESSION,
    callback: () => {
        if (Session.isAnonymousUser()) {
            return;
        }
        Navigation.navigate(ROUTES.HOME);
    }
});
```

I'm still not sure where that code should be called, and probably we'd need to check if the RHP is open (but probably not). The changes I'm doing in the server will insert the new authToken and delete the `authTokenType` property (so `Session.isAnonymousUser()` will return false as soon as the new data is inserted. I hope those changes will be live next week.

**Comment by marcochavezf** [comment]
One of two backend PRs was merged. The other one is still in review.

**Comment by marcochavezf** [comment]
The other PR was merged 🎉 we'll need to be deployed. I will post an update when the changes are live.

**Comment by marcochavezf** [comment]
Hi @daordonez11, the last [PR](https://github.com/Expensify/Web-Expensify/pull/37528) was deployed to staging, so the "claim" logic should be live now there. We're just re-using the `Session.signIn` action to merge the authenticated account into the anonymous account. The app sends under the hood the `authToken`, we verify it's an anonymous authToken and then do the merge.

I tested the changes with this [PR](https://github.com/Expensify/App/pull/23120), where I added some "dirty" changes based on yours. Feel free to update your PR using part of the new changes (it's ok to discard my PR since I only used it to test the backend changes quickly 😄).

**Comment by daordonez11** [comment]
Got it Marco thanks will continue working on this tomorrow then 🫡

**Comment by marcochavezf** [comment]
Cool, thanks @daordonez11! Any doubts/issues please let me know!

**Comment by daordonez11** [comment]
After this time merging was difficult so created new draft PR to continue work and fix issues!. Will update here when it is ready for review. PR: https://github.com/Expensify/App/pull/24083

**Comment by daordonez11** [comment]
Hey @marcochavezf  everything worked as expected! Already got the PR up and running you can give it a check and test already got an idea on how to hide sign in modal after command. Let me know what you think still missing 2 issues:
1. we're showing the scrollbar in the RHP (we should remove it)
    Regarding this one, the scrolbar exists in mobile and it is hidden, should we do something special in this modal case?
3. when the user clicks on one item from the FAB menu, the FAB menu is still open
    This one I will check tomorrow and comapre with other Modals, not sure why it is not working as is, maybe I'm missing a wrapper or something.

cc @mountiny

**Comment by marcochavezf** [comment]
Thanks for the update @daordonez11! Glad that this worked as expected 🎉 Also the condition to check if the sign-in modal route is active is a great idea!

> Regarding this one, the scrolbar exists in mobile and it is hidden, should we do something special in this modal case?

If it's hidden and if the user can't scroll I think it's fine

> when the user clicks on one item from the FAB menu, the FAB menu is still open

Can you record a video about this issue? Maybe we can do something similar to what you did [here](https://github.com/daordonez11/DanielExpensify/blob/8320e5df7248b2330a1fcdd3d6d687432d4a9572/src/libs/Navigation/AppNavigator/AuthScreens.js#L50-L52) to hide the FAB (I don't remember off the top of my head exactly how, but I think there's a function to close the FAB).

**Comment by daordonez11** [comment]
For sure @marcochavezf!

So regarding login my question is, in mWeb and native you can scroll and see the scrollbar even on first login, do we want to disable this in the modal? Only have the sign in form? If the answer is yes probably it's about using the prop and removing stuff.

Regarding FAB error it was reported by you [here](https://github.com/Expensify/App/issues/20558#issuecomment-1624450929) the error is that it is still opened when you click an action and the sign in modal is opened. I will review the flo, since the user is unauthenticated we might be missing some action of auto-hiding on element click.

**Comment by marcochavezf** [comment]
> So regarding login my question is, in mWeb and native you can scroll and see the scrollbar even on first login, do we want to disable this in the modal? Only have the sign in form? If the answer is yes probably it's about using the prop and removing stuff.

Ah yes, we just want to disable it in the modal, because in the footer we have other links and I think it could be a bit chaotic to handle the edge cases when someone opens some of those links in the RHP

> Regarding FAB error it was reported by you https://github.com/Expensify/App/issues/20558#issuecomment-1624450929 the error is that it is still opened when you click an action and the sign in modal is opened. I will review the flo, since the user is unauthenticated we might be missing some action of auto-hiding on element click.

Oh correct I forgot that issue, yeah I think alternatively we can hide it when the user user triggers the sign-in modal

**Comment by mountiny** [comment]
Amazing job @daordonez11! And I agree with Marco

**Comment by dylanexpensify** [comment]
@mollfpr let's get a review!

**Comment by dylanexpensify** [comment]
Bump @mollfpr

**Comment by daordonez11** [comment]
@dylanexpensify it was already approved by @mollfpr we are now waiting for final review from @marcochavezf

**Comment by dylanexpensify** [comment]
sorry for the miss, ty!

**Comment by daordonez11** [comment]
Hey guys! Thanks for the guidance and patience everyone. So finally seeing we got this through I wanted to open a discussion for reward on the issue. As discussed in this [slack thread](https://expensify.slack.com/archives/C01GTK53T8Q/p1686787152256199) the scope of the issue changed and there were some additional tasks done to achieve the final goal. For your analysis of the reward I wanted to list all the executed tasks in the issue:
- Approved proposal and [first PR](https://github.com/Expensify/App/pull/20733/files) got rejected due to the specificity of code for the specific issue.
    - Scope creep and we decide to handle it with navigation
- [First attempt PR](https://github.com/Expensify/App/pull/22005) was unifying the stack and discarded due to side effects found also seeing that it didn't fix the original issue.
- Second attempt and [final PR](https://github.com/Expensify/App/pull/24083) is Signin in RHP which was approved yesterday and included some improvements for AuthScreens

Finally thank you very much everyone for the opportunity and teachings it took some time but it is awesome to finally see this reaching main. And of course @mollfpr for the help with the final details to make it work smoothly!

cc @marcochavezf @mountiny @dylanexpensify

**Comment by marcochavezf** [comment]
Thanks @daordonez11 for your effort and I agree that this issue required exploration and back and forth discussion to reach to an optimal solution. Given the complexity of the task I think it would be ideal to double the bounty here. Thoughts @mountiny? (Since you were in the loop from the first attempt to solve the back navigation issue).

**Comment by dylanexpensify** [comment]
Waiting for regression period to pass, then will pay w/double bounty!

**Comment by daordonez11** [comment]
Hey there! @dylanexpensify this was deployed to prod on 23-08 I think regression period is already covered also I haven't seen any report yet 🫡.

Cc @marcochavezf @mountiny

**Comment by daordonez11** [comment]
> This seems to be ready for payment then https://github.com/Expensify/App/pull/24083
> 
> 
> 
> We should add a test for this for sure @daordonez11 can you suggest test steps please to cover this thoroughly?

Yeah for sure! I'll create something close to the BZ list and post it today

**Comment by daordonez11** [comment]
- [x] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
There is no offending PR this was created as sis since the definition of `AuthScreens`, the login page was just ina different `RootStack`
- [x] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:
    NA, this turned into a feature
- [x] Determine if we should create a regression test for this bug.
Here is the list of steps I would recommend:

1. Open a public room as unauthenticated user (i.e. https://staging.new.expensify.com/r/8813851443373010)
2. Click on sign-in
3. Click on the back button(Native or Header button), you should see the public room instead of closing the app
4. Enter sign-in in RHP and finish the sign-in flow with a new account
5. Open a new window in incognito mode, sign in with another account, and post something in the public room
6. Verify the message appears in the public room for the first user that claimed the anonymous account

 - Do we agree 👍 or 👎

**Comment by daordonez11** [comment]
Friendly bump @dylanexpensify

**Comment by daordonez11** [comment]
Hey guys, @marcochavezf @mountiny  do you know if maybe @dylanexpensify  is OOO? It's already been 2 weeks after the expected payment date.

**Comment by mountiny** [comment]
He was at the conference we were launching so I think he fell behind these chores @dylanexpensify could you look into these ones please?

**Comment by dylanexpensify** [comment]
Yes! On it - thanks for the ping!

**Comment by daordonez11** [comment]
Hey @dylanexpensify hope everything went well at the conference! Maybe a new task will be required since the original job is no longer available https://www.upwork.com/jobs/~017a21cdc86c5eee7f

**Comment by dylanexpensify** [comment]
Agree! On it now!

**Comment by dylanexpensify** [comment]
Payment summary:
- Contributor: @daordonez11 $2,000
- Contributor+: @mollfpr $2,000

[Upwork Job](https://www.upwork.com/jobs/~014d256c184b542ada)! Please apply!

**Comment by mollfpr** [comment]
@dylanexpensify Applied, thanks!

**Comment by daordonez11** [comment]
Applied too thanks @dylanexpensify

**Comment by dylanexpensify** [comment]
Woot!

**Comment by dylanexpensify** [comment]
payments sent! TY both!!
